gobblin-metrics/src/main/java/gobblin/metrics/MetricContext.java

```diff
@@ -0,0 +1,665 @@
+    return contextAwareMeter(name, ContextAwareMetricFactory.DEFAULT_CONTEXT_AWARE_METER_FACTORY);
+  }
+
+  /**
+   * Get a {@link ContextAwareMeter} with a given name.
+   *
+   * @param name name of the {@link ContextAwareMeter}
+   * @param factory a {@link ContextAwareMetricFactory} for building {@link ContextAwareMeter}s
+   * @return the {@link ContextAwareMeter} with the given name
+   */
+  public ContextAwareMeter contextAwareMeter(String name, ContextAwareMetricFactory<ContextAwareMeter> factory) {
+    return getOrCreate(name, factory);
+  }
+
+  /**
+   * Get a {@link ContextAwareHistogram} with a given name.
+   *
+   * @param name name of the {@link ContextAwareHistogram}
+   * @return the {@link ContextAwareHistogram} with the given name
+   */
+  public ContextAwareHistogram contextAwareHistogram(String name) {
+    return contextAwareHistogram(name, ContextAwareMetricFactory.DEFAULT_CONTEXT_AWARE_HISTOGRAM_FACTORY);
+  }
+
+  /**
+   * Get a {@link ContextAwareHistogram} with a given name.
+   *
+   * @param name name of the {@link ContextAwareHistogram}
+   * @param factory a {@link ContextAwareMetricFactory} for building {@link ContextAwareHistogram}s
+   * @return the {@link ContextAwareHistogram} with the given name
+   */
+  public ContextAwareHistogram contextAwareHistogram(String name,
+      ContextAwareMetricFactory<ContextAwareHistogram> factory) {
+    return getOrCreate(name, factory);
+  }
+
+  /**
+   * Get a {@link ContextAwareTimer} with a given name.
+   *
+   * @param name name of the {@link ContextAwareTimer}
+   * @return the {@link ContextAwareTimer} with the given name
+   */
+  public ContextAwareTimer contextAwareTimer(String name) {
+    return contextAwareTimer(name, ContextAwareMetricFactory.DEFAULT_CONTEXT_AWARE_TIMER_FACTORY);
+  }
+
+  /**
+   * Get a {@link ContextAwareTimer} with a given name.
+   *
+   * @param name name of the {@link ContextAwareTimer}
+   * @param factory a {@link ContextAwareMetricFactory} for building {@link ContextAwareTimer}s
+   * @return the {@link ContextAwareTimer} with the given name
+   */
+  public ContextAwareTimer contextAwareTimer(String name, ContextAwareMetricFactory<ContextAwareTimer> factory) {
+    return getOrCreate(name, factory);
+  }
+
+  /**
+   * Create a new {@link ContextAwareGauge} wrapping a given {@link com.codahale.metrics.Gauge}.
+   *
+   * @param name name of the {@link ContextAwareGauge}
+   * @param gauge the {@link com.codahale.metrics.Gauge} to be wrapped by the {@link ContextAwareGauge}
+   * @param <T> the type of the {@link ContextAwareGauge}'s value
+   * @return a new {@link ContextAwareGauge}
+   */
+  public <T> ContextAwareGauge<T> newContextAwareGauge(String name, Gauge<T> gauge) {
+    return new ContextAwareGauge<T>(this, name, gauge);
+  }
+
+  @Override
+  public boolean remove(String name) {
+    return this.contextAwareMetrics.remove(name) != null && super.remove(MetricRegistry.name(metricNamePrefix(), name));
```

Comment:
This will remove the metric from parent context regardless of whether other children of the parent context also have that metric. This is probably not the intended functionality.